File: importers/make_richmenu.swift
// Draw the LINE rich menu image. Run via importers/make_richmenu.py.
//
// Swift rather than Pillow because Pillow here is built without raqm, so it
// cannot shape Thai — vowels and tone marks land in the wrong place, which on
// a brand asset is worse than no asset. CoreText shapes Thai correctly, and
// /usr/bin/swift ships with the command line tools.
//
// LINE accepts 2500x1686 for a six-button menu; tap areas are declared
// separately in the JSON the Python side writes, in this same pixel space.

import Foundation
import AppKit

let W = 2500.0, H = 1686.0
let COLS = 3.0, ROWS = 2.0
let cw = W / COLS, ch = H / ROWS

struct Cell {
    let glyph: String
    let th: String
    let en: String
    let tint: NSColor
}

// Row one is why anybody opens it; row two is what we want them to do. Put the
// pleasure first — a person who came for today's fortune is already here when
// they notice their own shop needs a phone number.
let cells: [Cell] = [
    Cell(glyph: "🔮", th: "ดวงวันนี้", en: "Today's fortune",
         tint: NSColor(srgbRed: 0.76, green: 0.25, blue: 0.11, alpha: 1)),
    Cell(glyph: "🎋", th: "เซียมซี", en: "Fortune sticks",
         tint: NSColor(srgbRed: 0.62, green: 0.42, blue: 0.09, alpha: 1)),
    Cell(glyph: "🎪", th: "งานในเมือง", en: "What's on",
         tint: NSColor(srgbRed: 0.16, green: 0.47, blue: 0.84, alpha: 1)),
    Cell(glyph: "🏪", th: "ร้านนี้ของฉัน", en: "This is my shop",
         tint: NSColor(srgbRed: 0.56, green: 0.18, blue: 0.07, alpha: 1)),
    Cell(glyph: "📷", th: "ส่งรูป", en: "Send a photo",
         tint: NSColor(srgbRed: 0.18, green: 0.49, blue: 0.31, alpha: 1)),
    Cell(glyph: "✏️", th: "ตรงนี้ผิด", en: "Something's wrong",
         tint: NSColor(srgbRed: 0.35, green: 0.31, blue: 0.28, alpha: 1)),
]

// Draw into a bitmap of exactly these pixel dimensions. NSImage.lockFocus()
// picks up the screen's backing scale, which on a Retina Mac silently doubles
// the output to 5000x3372 — a size LINE rejects.
guard let rep = NSBitmapImageRep(
        bitmapDataPlanes: nil, pixelsWide: Int(W), pixelsHigh: Int(H),
        bitsPerSample: 8, samplesPerPixel: 4, hasAlpha: true, isPlanar: false,
        colorSpaceName: .deviceRGB, bytesPerRow: 0, bitsPerPixel: 0) else { exit(1) }
rep.size = NSSize(width: W, height: H)
guard let gctx = NSGraphicsContext(bitmapImageRep: rep) else { exit(1) }
NSGraphicsContext.saveGraphicsState()
NSGraphicsContext.current = gctx

// paper
NSColor(srgbRed: 0.984, green: 0.965, blue: 0.933, alpha: 1).setFill()
NSBezierPath(rect: NSRect(x: 0, y: 0, width: W, height: H)).fill()

func draw(_ s: String, font: NSFont, colour: NSColor, centreX: Double, baselineY: Double) {
    let attrs: [NSAttributedString.Key: Any] = [.font: font, .foregroundColor: colour]
    let a = NSAttributedString(string: s, attributes: attrs)
    let sz = a.size()
    a.draw(at: NSPoint(x: centreX - sz.width / 2, y: baselineY))
}

let thaiFont = NSFont(name: "Thonburi-Bold", size: 96)
    ?? NSFont.boldSystemFont(ofSize: 96)
let enFont = NSFont.systemFont(ofSize: 52, weight: .medium)
let glyphFont = NSFont(name: "AppleColorEmoji", size: 190)
    ?? NSFont.systemFont(ofSize: 190)

for (i, cell) in cells.enumerated() {
    let col = Double(i % Int(COLS)), row = Double(i / Int(COLS))
    // AppKit's origin is bottom-left; LINE counts rows from the top.
    let x = col * cw
    let y = H - (row + 1) * ch
    let inset = 18.0
    let r = NSRect(x: x + inset, y: y + inset, width: cw - inset * 2, height: ch - inset * 2)

    let card = NSBezierPath(roundedRect: r, xRadius: 46, yRadius: 46)
    NSColor.white.setFill(); card.fill()
    cell.tint.withAlphaComponent(0.9).setStroke(); card.lineWidth = 7; card.stroke()

    // a soft band of the tint at the top of each card, so the six read as a set
    let band = NSBezierPath(roundedRect: NSRect(x: r.minX, y: r.maxY - 130,
                                                width: r.width, height: 130),
                            xRadius: 46, yRadius: 46)
    cell.tint.withAlphaComponent(0.13).setFill(); band.fill()

    let cx = x + cw / 2
    draw(cell.glyph, font: glyphFont, colour: .black, centreX: cx, baselineY: y + ch * 0.52)
    draw(cell.th, font: thaiFont, colour: cell.tint, centreX: cx, baselineY: y + ch * 0.30)
    draw(cell.en, font: enFont, colour: NSColor(white: 0.45, alpha: 1),
         centreX: cx, baselineY: y + ch * 0.18)
}

NSGraphicsContext.restoreGraphicsState()

guard let png = rep.representation(using: .png, properties: [:]) else { exit(1) }
let out = CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : "richmenu.png"
try png.write(to: URL(fileURLWithPath: out))
print("wrote \(out)")

File: importers/make_richmenu.py
#!/usr/bin/env python3
"""Build the LINE rich menu — the image and the tap-area JSON that goes with it.

The rich menu is the grid pinned to the bottom of the LINE chat. It is the
closest thing this project has to a granny interface: six always-there buttons,
no typing, no account, inside the app she already has open. It costs nothing —
rich menus are on LINE's free plan.

Layout, deliberately:

    ดวงวันนี้      เซียมซี        งานในเมือง      <- why anyone opens it
    ร้านนี้ของฉัน   ส่งรูป         ตรงนี้ผิด        <- what we would like them to do

Pleasure on the top row. Somebody who came to shake the fortune sticks is
already here when they notice their own shop is missing a phone number.

    python3 importers/make_richmenu.py

Writes assets/richmenu.png and assets/richmenu.json. Upload both in LINE OA
Manager (Home -> Rich menus), or POST the JSON to the Messaging API if a
channel token exists. The image is drawn by make_richmenu.swift, because
Pillow here has no raqm and cannot shape Thai.
"""
import json
import os
import subprocess
import sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SWIFT = os.path.join(ROOT, "importers", "make_richmenu.swift")
PNG = os.path.join(ROOT, "assets", "richmenu.png")
JSON_OUT = os.path.join(ROOT, "assets", "richmenu.json")
SITE = "https://motdang.net/"

W, H, COLS, ROWS = 2500, 1686, 3, 2
CW, CH = W // COLS, H // ROWS

# Order must match the cells in make_richmenu.swift — same grid, same reading
# order, left to right then down.
ACTIONS = [
    {"type": "uri", "label": "duang", "uri": SITE + "widgets.html"},
    {"type": "uri", "label": "siamsi", "uri": SITE + "widgets.html#w-siamsi"},
    {"type": "uri", "label": "events", "uri": SITE + "events.html"},
    {"type": "uri", "label": "claim", "uri": SITE + "claim.html"},
    # These two are message actions on purpose. A photo is sent by sending a
    # photo — pushing someone to a web form to upload one is the barrier this
    # whole exercise exists to remove. Tapping puts the words in the box; she
    # attaches the picture and presses send, which she already knows how to do.
    {"type": "message", "label": "photo", "text": "ส่งรูป"},
    {"type": "message", "label": "fix", "text": "ตรงนี้ผิด"},
]

AREAS = []
for i, action in enumerate(ACTIONS):
    col, row = i % COLS, i // COLS
    AREAS.append({
        "bounds": {"x": col * CW, "y": row * CH, "width": CW, "height": CH},
        "action": action,
    })

MENU = {
    "size": {"width": W, "height": H},
    "selected": True,
    "name": "Mot Dang - main",
    "chatBarText": "เมนูมดแดง",
    "areas": AREAS,
}


def main():
    os.makedirs(os.path.join(ROOT, "assets"), exist_ok=True)
    try:
        out = subprocess.run(["swift", SWIFT, PNG], capture_output=True, text=True, timeout=300)
    except FileNotFoundError:
        raise SystemExit("swift not on PATH — install the Xcode command line tools")
    if out.returncode != 0:
        raise SystemExit("swift failed:\n" + (out.stderr or out.stdout)[:2000])

    size = os.path.getsize(PNG)
    with open(JSON_OUT, "w") as fh:
        json.dump(MENU, fh, ensure_ascii=False, indent=1)

    print(f"🐜 {PNG} ({size:,} bytes, {W}x{H})")
    print(f"🐜 {JSON_OUT} — {len(AREAS)} tap areas")
    if size > 1_000_000:
        print("⚠  over LINE's 1 MB limit — reduce detail or save as JPEG")
    for a in AREAS:
        act = a["action"]
        where = act.get("uri") or f'sends "{act.get("text")}"'
        b = a["bounds"]
        print(f'   ({b["x"]:>4},{b["y"]:>4}) {b["width"]}x{b["height"]}  {act["label"]:<7} -> {where}')
    print("\nUpload both in LINE OA Manager -> Home -> Rich menus.")
    print("The two message buttons need a matching auto-reply; see the keywords")
    print("printed by --replies.")
    if "--replies" in sys.argv:
        print("\n--- suggested auto-replies (OA Manager -> Auto-response) ---")
        print('keyword "ส่งรูป":')
        print("  ส่งรูปมาได้เลยเจ้า ถ่ายจากมือถือก็ได้ ใส่ชื่อร้านมาด้วยนะ")
        print("  แล้วเราจะลงเครดิตชื่อคุณไว้ใต้รูป")
        print("  (Send the photo right here. Tell us which place it is, and your")
        print("   name goes under the picture.)")
        print('\nkeyword "ตรงนี้ผิด":')
        print("  บอกมาสั้นๆ ได้เลย ว่าตรงไหนผิด แล้วที่ถูกคืออะไร")
        print("  ถ้ามีลิงก์หน้าร้านด้วยยิ่งดี")
        print("  (Tell us briefly what is wrong and what it should be. A link to")
        print("   the page helps.)")


if __name__ == "__main__":
    main()
